modules/cluster-management/src/main/java/org/apache/ignite/internal/cluster/management/ClusterManagementGroupManager.java

```diff
@@ -434,6 +441,12 @@ private void handleCancelInit(CancelInitMessage msg) {
      * Completely destroys the local CMG Raft service.
      */
     private void destroyCmg() {
+        try {
+            fireEvent(ClusterManagerGroupEvent.BEFORE_RAFT_GROUP_DESTROY, EmptyEventParams.INSTANCE).join();
```

Comment:
It's better to use `get` with a timeout. Also, this method is called inside a network thread, we should probably dispatch it onto a thread pool or make it asynchronous.

Reply:
How about not using get/join at all? Just collect the listeners' futures to one using `CompletableFuture#allOf()` and than on that aggregated future make `.thenRun()` doing the rest of the method; `destroyCmg()` would return a `CompletableFuture` as well; the code that calls it and needs to execute some logic strictly after the CMG has been destroyed can just chain on the future it returns.

Reply:
Currently, there's no code after `destoryCmg` just the error handling message and the busyLock unlock.

Reply:
In `handleClusterState()`, there is a call to `destroyCmg()`, and there is a `initCmgRaftService(state)` call that probably relies on the CMG being destroyed at the moment when it's made.

Reply:
Yeah, I saw that afterwards. Missed that one on the first iteration.